src/org/sosy_lab/common/collect/OrderStatisticMap.java

```diff
@@ -0,0 +1,138 @@
+/*
+ *  SoSy-Lab Common is a library of useful utilities.
+ *  This file is part of SoSy-Lab Common.
+ *
+ *  Copyright (C) 2007-2017  Dirk Beyer
+ *  All rights reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.sosy_lab.common.collect;
+
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.NavigableMap;
+
+public interface OrderStatisticMap<K, V> extends NavigableMap<K, V> {
+
+  /**
+   * Returns the element of this set with the given rank. The lowest element in the set has rank ==
+   * 0, the largest element in the set has rank == size - 1.
+   *
+   * @param pIndex the rank of the element to return
+   * @return the element of this set with the given rank
+   * @throws IndexOutOfBoundsException if the given rank is out of the range of this set (i.e.,
+   *     pRank &lt; 0 || pRank &gt;= size)
+   */
+  K getKeyByRank(int pIndex);
```

Comment:
Is this operation common enough to have its own method?
If yes, it could be a default method delegating to `getEntryByRank`, otherwise we could leave it out.

Reply:
I wanted to provide it in reference to `NavigableMap`.
It also includes methods `lower*`, `floor*`, etc. for `Entry` and `Key`.
But currently I'm not using or needing it.